Yes, makes sense. I should have included something like this when I added the json tests.

I think https://github.com/bitcoin/bitcoin/pull/9023 (which outputs a contextual diff for failing test cases) should make it much easier to track down formatting errors, since it'll point you to the exact line that's different.

However, I think this is still useful for verifying that the output is indeed valid json.

Only nit is that `compare_output()` can raise errors which aren't caught anywhere and don't result in useful logging. For example, here's the logging if the output file is invalid json:

```
~/bitcoin/src$ test/bitcoin-util-test.py --srcdir='.'
Traceback (most recent call last):
  File "test/bitcoin-util-test.py", line 32, in <module>
    bctest.bctester(srcdir + "/test/data", "bitcoin-util-test.json", buildenv, verbose = verbose)
  File "/home/vagrant/bitcoin/src/test/bctest.py", line 72, in bctester
    bctest(testDir, testObj, buildenv.exeext)
  File "/home/vagrant/bitcoin/src/test/bctest.py", line 50, in bctest
    if not compare_output(outs[0], outputData, outputType):
  File "/home/vagrant/bitcoin/src/test/bctest.py", line 14, in compare_output
    match = (json.loads(a) == json.loads(b))
  File "/usr/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 382, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Invalid control character at: line 3 column 14 (char 95)
```

It'd be nice to wrap the json calls in try/except, and maybe have the final else branch of `compare_output()` print an error and exit rather than raise an error.

--- a/src/test/bctest.py
+++ b/src/test/bctest.py
@@ -6,6 +6,15 @@ import subprocess
 import os
 import json
 import sys
+import binascii
+
+def parse_output(a, fmt):
+	if fmt == 'json': # json: compare parsed data
+		return json.loads(a)
+	elif fmt == 'hex': # hex: parse and compare binary data
+		return binascii.a2b_hex(a.strip())
+	else:
+		raise NotImplementedError("Don't know how to compare %s" % fmt)
 
 def bctest(testDir, testObj, exeext):
 
@@ -23,6 +32,7 @@ def bctest(testDir, testObj, exeext):
 	outputData = None
 	if "output_cmp" in testObj:
 		outputFn = testObj['output_cmp']
+		outputType = os.path.splitext(outputFn)[1][1:] # output type from file extension (determines how to compare)
 		outputData = open(testDir + "/" + outputFn).read()
 		if not outputData:
 			print("Output data missing for " + outputFn)
@@ -34,9 +44,23 @@ def bctest(testDir, testObj, exeext):
 		print("OSError, Failed to execute " + execprog)
 		sys.exit(1)
 
-	if outputData and (outs[0] != outputData):
-		print("Output data mismatch for " + outputFn)
-		sys.exit(1)
+	if outputData:
+		try:
+			a_parsed = parse_output(outs[0], outputType)
+		except Exception as e:
+			print('Error parsing command output as %s: %s' % (outputType,e))
+			sys.exit(1)
+		try:
+			b_parsed = parse_output(outputData, outputType)
+		except Exception as e:
+			print('Error parsing expected output %s as %s: %s' % (outputFn,outputType,e))
+			sys.exit(1)
+		if a_parsed != b_parsed:
+			print("Output data mismatch for " + outputFn + " (format " + outputType + ")")
+			sys.exit(1)
+		if outs[0] != outputData:
+			print("Output formatting mismatch for " + outputFn + " (format " + outputType + ")")
+			sys.exit(1)
 
 	wantRC = 0
 	if "return_code" in testObj:

